src/main/java/de/thomas_oster/liblasercut/drivers/ThunderLaser.java

```diff
@@ -0,0 +1,1020 @@
+  protected static final String SETTING_BED_HEIGHT = "Bed height (mm)";
+  protected static final String SETTING_RASTER_WHITESPACE = "Additional space per Raster line (mm)";
+  protected static final String SETTING_USE_BIDIRECTIONAL_RASTERING = "Use bidirectional rastering";
+  // config values
+  private static final long[] JogAcceleration = {200000,50000,600000};
+  private static final long[] JogMaxVelocity = {16,16,2048};
+  private static final long[] EngraveAcceleration = {200000,50000,600000};
+  private static final long[] EngraveMaxVelocity = {800,800,2048};
+  private static final long[] VectorAcceleration = {100000,25000,20000};
+  private static final long[] VectorMaxVelocity = {1000,1000,1000};
+  private static final byte FlipLaserPWMPower = 1;
+  private static final byte FlipLaserOutput = 0;
+
+  private static final byte HomeDirection = 1;
+  private static final byte[] FlipHomeDirection = {1,0,0};
+  private static final byte[] LimitContCondition = {0,0,0,0};
+  private static final long[] MaxSteps = {250,500,500};
+  private static final long[] TableSize = {20000,12000,30000};
+
+  private int mm2focus(float mm)
+  {
+    return (int) (mm / FOCUSWIDTH);
+  }
+
+  private float focus2mm(int focus)
+  {
+    return (float) (focus * FOCUSWIDTH);
+  }
+
+  public ThunderLaser()
+  {
+    System.out.println("ThunderLaser() constructor");
+  }
+
+  @Override
+  public ThunderLaserProperty getLaserPropertyForVectorPart()
+  {
+    return new ThunderLaserProperty();
+  }
+
+  @Override
+  public ThunderLaserProperty getLaserPropertyForRasterPart()
+  {
+    return new ThunderLaserProperty();
+  }
+
+  @Override
+  public ThunderLaserProperty getLaserPropertyForRaster3dPart()
+  {
+    return new ThunderLaserProperty();
+  }
+
+  @Override
+  public boolean canEstimateJobDuration()
+  {
+    return true;
+  }
+
+  /**
+   * When rastering, whether to always cut from left to right, or to cut in both
+   * directions? (i.e. use the return stroke to raster as well)
+   */
+  protected boolean useBidirectionalRastering = false;
+
+  public boolean getUseBidirectionalRastering()
+  {
+    return useBidirectionalRastering;
+  }
+
+  public void setUseBidirectionalRastering(boolean useBidirectionalRastering)
+  {
+    this.useBidirectionalRastering = useBidirectionalRastering;
+  }
+
+  /**
+   * 'runway' for laser to get up to speed when rastering
+   *
+   */
+  private double addSpacePerRasterLine = 2;
+
+  /**
+   * Get the value of addSpacePerRasterLine
+   *
+   * @return the value of addSpacePerRasterLine
+   */
+  public double getAddSpacePerRasterLine() {
+    return addSpacePerRasterLine;
+  }
+
+  /**
+   * Set the value of addSpacePerRasterLine
+   *
+   * @param addSpacePerRasterLine new value of addSpacePerRasterLine
+   */
+  public void setAddSpacePerRasterLine(double addSpacePerRasterLine) {
+    this.addSpacePerRasterLine = addSpacePerRasterLine;
+  }
+  /*
+   * estimateJobDuration - copied from EpilogCutter
```

Comment:
Did you change anything except for the constants? If no, it's probably better to move this method to something like LaserCutter.estimateJobDuration(job, vector_movespeed_x, vector_movespeed_y, vector_lineoffset, raster_lineoffset, ...) and just call it from ThunderLaser and from EpilogCutter with the appropriate constants.

Reply:
Yes, once this function is in a useful state (hint: it's not ;-))

Reply:
see my ruida-help branch, where it has been outsourced after some refactoring.

Reply:
Sorry Max, your ruida-help branch is (at least) 2 levels above my Java capabilities :-(

Reply:
If you're stuck, just ping me as soon as everything else is finished and I'll try to do the rest (deduplicating time estimation, switch to integer-based property class).

Reply:
will be fixed by https://github.com/kkaempf/LibLaserCut/pull/1, you just need to calibrate the constants (engrave/cutting/move speed at 100%)